WordPress/Classes/ViewRelated/Blog/Site Picker/BlogList/BlogListViewModel.swift

```diff
@@ -0,0 +1,107 @@
+import SwiftUI
+
+final class BlogListViewModel: ObservableObject {
+    @Published var recentSites: [BlogListView.Site]
+    @Published var pinnedSites: [BlogListView.Site]
+    @Published var allRemainingSites: [BlogListView.Site]
+    @Published var searchSites: [BlogListView.Site]
+
+    private let contextManager: ContextManager
+    private let dataSource: BlogListDataSource
+
+    init(contextManager: ContextManager = ContextManager.sharedInstance()) {
+        self.contextManager = contextManager
+        self.dataSource = Self.createDataSource(contextManager: contextManager)
+        pinnedSites = Self.filteredPinnedSites(allBlogs: dataSource.filteredBlogs)
+        recentSites = Self.filteredRecentSites(allBlogs: dataSource.filteredBlogs)
+        allRemainingSites = Self.filteredAllRemainingSites(allBlogs: dataSource.filteredBlogs)
+        searchSites = dataSource.filteredBlogs.compactMap(BlogListView.Site.init)
+    }
+
+    private static func createDataSource(contextManager: ContextManager) -> BlogListDataSource {
+        // Utilize existing DataSource class to fetch blogs.
+        let config = BlogListConfiguration.defaultConfig
+        let dataSource = BlogListDataSource(contextManager: contextManager)
+        dataSource.shouldHideSelfHostedSites = config.shouldHideSelfHostedSites
+        dataSource.shouldHideBlogsNotSupportingDomains = config.shouldHideBlogsNotSupportingDomains
+        return dataSource
+    }
+
+    var allBlogs: [Blog] {
+        return dataSource.filteredBlogs
+    }
+
+    func updateSearchText(_ newText: String) {
+        if newText.isEmpty {
+            searchSites = allBlogs.compactMap(BlogListView.Site.init)
+        } else {
+            searchSites = allBlogs
+                .filter {
+                    $0.url?.lowercased().contains(newText.lowercased()) == true
+                    || $0.title?.lowercased().contains(newText.lowercased()) == true
+                }
+                .compactMap(BlogListView.Site.init)
+        }
+    }
+
+    func togglePinnedSite(siteID: NSNumber?) {
+        guard let blog = allBlogs.first(where: { $0.dotComID == siteID }) else {
+            return
+        }
+
+        blog.pinnedDate = blog.pinnedDate == nil ? Date() : nil
+        pinnedSites = Self.filteredPinnedSites(allBlogs: allBlogs)
+
+        let beforeRecentsCount = recentSites.count
+        recentSites = Self.filteredRecentSites(allBlogs: allBlogs)
+
+        if recentSites.count == beforeRecentsCount {
+            allRemainingSites = Self.filteredAllRemainingSites(allBlogs: allBlogs)
+        }
+        contextManager.saveContextAndWait(contextManager.mainContext)
```

Comment:
Why are we using `saveContextAndWait` instead of `save`?

Reply:
Resolved in [5942cf7](https://github.com/wordpress-mobile/WordPress-iOS/pull/22833/commits/5942cf79ae1a48501880961001b389939406ffe7)

Reply:
Test fail due to save being asynchronous. Do you know how to wait for the `save` / `perform` from the unit tests?
I tried using `expectation(forNotification: .NSManagedObjectContextDidSave, object: mainContext)` and waiting for the expectation but it doesn't seem to work. The functions that use that expectation also seem to use `saveContextAndWait`

Reply:
Actually, when I call `save`, the list does not animate on change anymore. Perhaps that's the reason there are more calls to `savenContextAndWait` from the production code. I don't think we can call `save` here.

Reply:
Thanks for sharing the context!

> Test fail due to save being asynchronous. Do you know how to wait for the save / perform from the unit tests?

We can do this by injecting a completion handler into the function and calling it when the save is complete using `save(_ context: NSManagedObjectContext, completion: (() -> Void)?, on queue: DispatchQueue)`

> Actually, when I call save, the list does not animate on change anymore.

Interesting 🤔  I don't fully understand why. The pro of using save is that it should offer a snappier experience for the user. Since we don't need to wait for the save, we can proceed with animating the change and unblocking the UI. However, this causing the animation to stop working is not great. It's a small thing, and I am fine with not spending more time investigating it 👍